src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp

```diff
@@ -0,0 +1,363 @@
+  regularization = 0.5 * lambda * arma::dot(parameters,
+      parameters);
+
+  cost = loss + regularization;
+  return cost;
+}
+
+template <typename MatType>
+template <typename GradType>
+void LinearSVMFunction<MatType>::Gradient(
+    const arma::mat& parameters,
+    GradType& gradient)
+{
+  arma::mat scores = parameters * dataset;
+  arma::mat correctScores = arma::repmat(arma::ones(numClasses).t()
+      * (scores % groundTruth), numClasses, 1);
+  arma::mat margin = scores - correctScores + delta
+      - (delta * groundTruth);
+
+  // For each sample, find the total number of classes where
+  // ( margin > 0 ).
+  arma::mat mask = margin.for_each([](arma::mat::elem_type& val)
+      { val = (val > 0) ? 1: 0; });
+
+  arma::mat difference = groundTruth
+      % (-arma::repmat(arma::sum(mask), numClasses, 1)) + mask;
+
+  gradient = difference * dataset.t();
+  gradient /= dataset.n_cols;
+
+  // Adding the regularization contribution to the gradient.
+  gradient += lambda * parameters;
+}
+
+template <typename MatType>
+template <typename GradType>
+void LinearSVMFunction<MatType>::Gradient(
+    const arma::mat& parameters,
+    const size_t firstId,
+    GradType& gradient,
+    const size_t batchSize)
+{
+  const size_t lastId = firstId + batchSize - 1;
+  arma::mat scores = parameters * dataset;
+  arma::mat correctScores = arma::repmat(arma::ones(numClasses).t()
+      * (scores % groundTruth), numClasses, 1);
+  arma::mat margin = scores - correctScores + delta
+      - (delta * groundTruth);
+
+  // For each sample, find the total number of classes where
+  // ( margin > 0 ).
+  arma::mat mask = margin.for_each([](arma::mat::elem_type& val)
+      { val = (val > 0) ? 1: 0; });
+
+  arma::mat difference = groundTruth
+      % (-arma::repmat(arma::sum(mask), numClasses, 1)) + mask;
+
+  gradient = difference.cols(firstId, lastId)
+      * dataset.cols(firstId, lastId).t();
+  gradient /= batchSize;
+
+  // Adding the regularization contribution to the gradient.
+  gradient += lambda * parameters;
+}
+
+template <typename MatType>
+template <typename GradType>
+double LinearSVMFunction<MatType>::EvaluateWithGradient(
+    const arma::mat& parameters,
+    GradType& gradient) const
+{
+  double loss, regularization, cost;
+
+  arma::mat scores = parameters * dataset;
+  arma::mat correctScores = arma::repmat(arma::ones(numClasses).t()
+      * (scores % groundTruth), numClasses, 1);
+  arma::mat margin = scores - correctScores + delta
+      - (delta * groundTruth);
+
+  // For each sample, find the total number of classes where
+  // ( margin > 0 ).
+  arma::mat mask = margin.for_each([](arma::mat::elem_type& val)
+      { val = (val > 0) ? 1: 0; });
+
+  arma::mat difference = groundTruth
+      % (-arma::repmat(arma::sum(mask), numClasses, 1)) + mask;
+
+  gradient = difference * dataset.t();
+  gradient /= dataset.n_cols;
+
+  // Adding the regularization contribution to the gradient.
+  gradient += lambda * parameters;
```

Comment:
Some optimizers, like Hogwild!, depend on the gradient actually being sparse: http://ensmallen.org/docs.html#sparse-differentiable-separable-functions.  But in this case, whenever `lambda > 0`, the gradient won't be sparse.  I wonder if this means `LinearSVMFunction` is not a great candidate for parallelization in that manner.

(However, it doesn't really matter for your implementation.  I am just pointing it out.)

Reply:
In the latest commits (after your last evaluation), I've added `fitIntercept` and one of the the function I'm using is `.insert_rows` for the evaluation of `Gradient`. Since `.insert_rows` is not a member function of `arma::sp_mat`, as of now, we cannot use `Hogwild` or any other optimizer that would require gradient to be `sparse`.

I'll still be looking for alternate ways so as we can use Gradient matrix as `arma::sp_mat`.

Reply:
Yeah, we should avoid `insert_rows()`.  In fact we should avoid that anyway because it will copy the original `gradient` matrix into a newly allocated larger one, which is inefficient.  Try allocating `gradient` with the correct size once and then using, e.g., `gradient.submat(a, b, c, d) = dataset * difference.t()` (you'll have to compute what a, b, c, and d should be).

Reply:
So for this, we would have to allocate size for `Gradient` so that we can add that `submat` and then the gradient of `bias terms` in the last row. Am I right?

Reply:
Yes, you can do that, and you can set the size with `gradient.set_size()`.